week5/Q1.py

```diff
@@ -0,0 +1,149 @@
+class OrientedGraph:
+	'''
+	Implementation of oriented graph. All of the nodes have unique values.
+
+	Fields:
+		nodes: set of Nodes, all nodes of the graph
+	'''
+	def __init__(self):
+		self.nodes = set()
+
+	def add(self, value, parent_value = None):
+		'''
+		Adds new Node to the OrientedGraph or updates existing ones.
+
+		Args:
+			value: str
+			parent_value: str
+		'''
+		node = None
+		parent = None
+		for inode in self.nodes:
+			if inode.value == value:
+				node = inode
+			if inode.value == parent_value:
+				parent = inode
+
+		if parent_value:
+			if node:
+				node.parents = node.parents.union({parent})
+				parent.children = parent.children.union({node})
+			else:
+				node = Node(value, parents = {parent})
+				parent.children = parent.children.union({node})
+				self.nodes.add(node)
+		else:
+			if not node:
+				node = Node(value)
+				self.nodes.add(node)
+
+
+class Node:
+	'''
+	Element of OrientedGraph.
+
+	Fields:
+		value: str
+		parents: set of Nodes, from which Node has incoming edges
+		children: set of Nodes, to which Node has outcoming edges
+	'''
+	def __init__(self, value = None, parents = set(), children = set()):
+		self.value = value
+		self.parents = parents
+		self.children = children
+
+
+def _drop_empty_words(words):
+	'''
+	Removes empty elements from list.
+
+	Args:
+		words: list
+
+	Returns:
+		list
+	'''
+	words = [word for word in words if word != '']
+	return words
+
+
+def _get_ordered_letters(dictionary, letter_graph):
```

Comment:
The `get` part of the name is a bit misleading, since this actually adds edges to an existing graph.
Maybe `_add_ordered_letters` would be a clearer name. Also, since this actually modified the given graph, you do not need to return it as well (it might be unclear whether it is modified or copied if you return it).